### PR TITLE
[4.0] Fix Untranslated tip

### PR DESCRIPTION
**Pull request by infograf768**
### Summary of Changes 

As title says. The tip displays in the articles Manager when an article is Unpublished.
It concerns the `COM_CONTENT_CHANGE_STAGE` lang string

### Testing Instructions
Create an article as Unpublished. 
make sure publish up and down are empty
Hover the Status icon.

### Before patch

Screen Shot 2019-11-28 at 11 06 52

### After patch
Screen Shot 2019-11-28 at 11 05 09

### About this patch
It is the simplest solution but we have another solution by passing through Text the string in both 
`articles/default.php` and `featured/default.php`

```
													->addState(ContentComponent::CONDITION_PUBLISHED, '', 'publish', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JPUBLISHED'])
													->addState(ContentComponent::CONDITION_UNPUBLISHED, '', 'unpublish', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JUNPUBLISHED'])
													->addState(ContentComponent::CONDITION_ARCHIVED, '', 'archive', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JARCHIVED'])
													->addState(ContentComponent::CONDITION_TRASHED, '', 'trash', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JTRASHED'])
													
```
I can change the PR if judged necessary.

### CODE REVIEW

**Comment by SharkyKZ** [comment]
>another solution by passing through Text the string in both
articles/default.php and featured/default.php

That's better. But I wonder whether passed strings should be translated or not. Here they're not translated:

https://github.com/joomla/joomla-cms/blob/551e3569883e9ffc1ab8784ddeeb6874e94087c6/libraries/src/Button/PublishedButton.php#L32-L35

So maybe translation should be performed in `Joomla\CMS\Button\PublishedButton::render()` instead.

**Comment by infograf768** [comment]
Afaik that part of PublishButton is not used for articles but may be for other managers and it is translated somewhere else. The articles manager loads a new PublishButton with specific states because of workflow. Then it is displayed by the transition-button where the textip (data content) depends of the state of the article obtained by Publish button when the article is published.
When it is unpublished, we need to use this pr or add Text:: in the default.php.
So we have to decide which solution to use OR refactor the whole stuff.

**Comment by SharkyKZ** [comment]
It's not translated as far as I can tell. After removing custom `com_content` states, I get `JLIB_HTML_PUBLISH_ITEM` in the tooltip which is what other components will get with default button.

**Comment by infograf768** [comment]
> It's not translated as far as I can tell. 

Indeed. In fact it looks like it is never used in core... What is used elsewhere is JGrid.
Only occurence of `new PublishedButton` are in both `articles/default.php` and `featured/default.php`

Looks like it "could" be used by 3rd party in the future.

Therefore I think it is safer to add the `Text::` both in the default.php files and PublishedButton itself. On it.

**Comment by infograf768** [comment]
Please test again.

**Comment by SharkyKZ** [comment]
Grrrr... Now it looks weird because title has to be translated but tip_title doesn't 🤷‍♂.

**Comment by infograf768** [comment]
The responsible for this mess is https://github.com/joomla/joomla-cms/pull/18501

**Comment by infograf768** [comment]
@SharkyKZ 
I found where the original issue comes from.
It was my patch https://github.com/joomla/joomla-cms/pull/27078 ;)
I had modified the transition-button there to prevent double translation when we had "Start Date etc.)
Debug Lang was marking it this way:
Screen Shot 2019-11-30 at 12 22 40

I patched this way
```
- data-content="<?php echo HTMLHelper::_('tooltipText', Text::_($title), '', 0); ?>"
+ data-content="<?php echo HTMLHelper::_('tooltipText', $title, '', 0); ?>"
```

Therefore the $title is untranslated when the article is unpublished.

We have to take a decision. Do we reinstate the Text there (and get a wrong debug lang) or modify as done in this PR?

**Comment by SharkyKZ** [comment]
I just don't like that one string we pass has to be translated but the other doesn't. I see two options here:

1. Remove translation from button layout and pass only translated strings to `PublishedButton::addState()`.

2. Remove translation from button layout and perform all translations in `PublishedButton::render()`.

**Comment by infograf768** [comment]
I'm afraid I can't do what you suggest.

**Comment by SharkyKZ** [comment]
You mean you don't know how to do it or you don't agree with proposed solutions?

**Comment by infograf768** [comment]
Don’t know how. 😄

**Comment by SharkyKZ** [comment]
Remove `Text::_()` calls from `layouts/joomla/button/transition-button.php` and add them to tip_title constants like you already did for title, e.g. here https://github.com/joomla/joomla-cms/blob/b264d75b2399963f622ac640e2221cbae1f21581/administrator/components/com_content/tmpl/articles/default.php#L252

This should also fix remaining double translations which aren't visible in core.

**Comment by infograf768** [comment]
@SharkyKZ 
Done and it does work indeed (changes were also needed in PublishedButton).

BUT, remains a case.
See transition-button `$only_icon = empty($options['transitions']);`
https://github.com/joomla/joomla-cms/blob/4.0-dev/layouts/joomla/button/transition-button.php#L31-L40

I did not touch for the moment at the `Text::_()` stuff when `$only_icon` is set as I don't know how to test that part. The logic is to also modify there but not sure at all.

If you don't know, maybe @bembelimen knows the uses of that part.

**Comment by SharkyKZ** [comment]
Yes, remove those `Text::_()` calls too. It works the same way. For testing purposes you can just set `$only_icon` to `true`.

**Comment by infograf768** [comment]
Done. Can be now tested. I fail to understand the use of `$only_icon`. Why would anyone lose the ability to modify via the icon is a mystery to me...

**Comment by SharkyKZ** [comment]
Remove `use Joomla\CMS\Language\Text;` from button layout please.

**Comment by infograf768** [comment]
Done.

File: libraries/src/Button/PublishedButton.php [hunk 1]
```diff
@@ -29,10 +29,10 @@ class PublishedButton extends ActionButton
...
 	protected function preprocess()
 	{
-		$this->addState(1, 'unpublish', 'publish', 'JLIB_HTML_UNPUBLISH_ITEM', ['tip_title' => 'JPUBLISHED']);
-		$this->addState(0, 'publish', 'unpublish', 'JLIB_HTML_PUBLISH_ITEM', ['tip_title' => 'JUNPUBLISHED']);
-		$this->addState(2, 'unpublish', 'archive', 'JLIB_HTML_UNPUBLISH_ITEM', ['tip_title' => 'JARCHIVED']);
-		$this->addState(-2, 'publish', 'trash', 'JLIB_HTML_PUBLISH_ITEM', ['tip_title' => 'JTRASHED']);
+		$this->addState(1, 'unpublish', 'publish', Text::_('JLIB_HTML_UNPUBLISH_ITEM'), ['tip_title' => 'JPUBLISHED']);
+		$this->addState(0, 'publish', 'unpublish', Text::_('JLIB_HTML_PUBLISH_ITEM'), ['tip_title' => 'JUNPUBLISHED']);
+		$this->addState(2, 'unpublish', 'archive', Text::_('JLIB_HTML_UNPUBLISH_ITEM'), ['tip_title' => 'JARCHIVED']);
+		$this->addState(-2, 'publish', 'trash', Text::_('JLIB_HTML_PUBLISH_ITEM'), ['tip_title' => 'JTRASHED']);
```
Comment (SharkyKZ) [inline libraries/src/Button/PublishedButton.php]: Missing tip_title translations.

File: layouts/joomla/button/transition-button.php [hunk 2]
```diff
@@ -31,8 +30,8 @@
 <?php if ($only_icon) : ?>
 	<span class="tbody-icon mr-1">
 		<span class="<?php echo $this->escape($icon ?? ''); ?> <?php echo $tip ? 'hasPopover' : ''; ?>"
-			title="<?php echo HTMLHelper::_('tooltipText', Text::_($tipTitle ? : $title), '', 0); ?>"
-			data-content="<?php echo HTMLHelper::_('tooltipText', Text::_($title), '', 0); ?>"
+			title="<?php echo HTMLHelper::_('tooltipText', $tipTitle ? : $title, '', 0); ?>"
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]:
```suggestion
			title="<?php echo HTMLHelper::_('tooltipText', $tipTitle ?: $title, '', 0); ?>"
```

File: layouts/joomla/button/transition-button.php [hunk 3]
```diff
@@ -43,7 +42,7 @@
 			href="javascript://"
 		<?php endif; ?>
 
-		title="<?php echo HTMLHelper::_('tooltipText', Text::_($tipTitle ? : $title), '', 0); ?>"
+		title="<?php echo HTMLHelper::_('tooltipText', $tipTitle ? : $title, '', 0); ?>"
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]:
```suggestion
		title="<?php echo HTMLHelper::_('tooltipText', $tipTitle ?: $title, '', 0); ?>"
```

File: libraries/src/Button/PublishedButton.php (hunk 1 above)
Comment (infograf768) [reply libraries/src/Button/PublishedButton.php]: did it, but not sure. Maybe I should rather have taken off translation for the JLIB ones.
See FeaturedButton which works perfectly without translate for these in preprocess()

File: layouts/joomla/button/transition-button.php (hunk 2 above)
Comment (infograf768) [reply layouts/joomla/button/transition-button.php]: done

File: layouts/joomla/button/transition-button.php (hunk 3 above)
Comment (infograf768) [reply layouts/joomla/button/transition-button.php]: done

**Comment by infograf768** [comment]
needs more work for featured.
On it

**Comment by SharkyKZ** [comment]
Remove `Text::_()` here too please: https://github.com/joomla/joomla-cms/blob/0f6751fc049bd9dc0705e62286b18e552eda444b/layouts/joomla/button/action-button.php#L40

And with that `use Joomla\CMS\Language\Text;` too.

**Comment by infograf768** [comment]
LOL
Hope it was the last one...

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 0318b14ddfb31ca27ecdeedb8e6e78229f6889d4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27174">issues.joomla.org/tracker/joomla-cms/27174</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 0318b14ddfb31ca27ecdeedb8e6e78229f6889d4<br /><br />RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27174">issues.joomla.org/tracker/joomla-cms/27174</a>.</sub>

**Comment by wilsonge** [comment]
This looks better to me translating the tips this way (sorry for not looking before JM!). Thanks!

**Comment by infograf768** [comment]
Thanks to all who helped on this. Was a long road... ;)
I still have no idea why we offer the variable `$only_icon` though except for overrides of the admin templates.

**Pull request merged**